charms/tensorboard-controller/src/charm.py

```diff
@@ -1,184 +1,302 @@
+
+    def _on_install(self, _) -> None:
+        """Handle install event."""
+        # deploy K8s resources to speed up deployment
+        self._apply_k8s_resources()
+
+    def _on_upgrade(self, _) -> None:
+        """Handle upgrade event."""
+        # force conflict resolution in K8s resources update
+        self._on_event(_, force_conflicts=True)
+
+    def _on_remove(self, _) -> None:
+        """Handle remove event."""
+        delete_error = None
+        self.unit.status = MaintenanceStatus("Removing K8s resources")
+        k8s_resources_manifests = self.k8s_resource_handler.render_manifests()
+        crd_resources_manifests = self.crd_resource_handler.render_manifests()
+        try:
+            delete_many(self.k8s_resource_handler.lightkube_client, k8s_resources_manifests)
+        except ApiError as error:
+            # do not log/report when resources were not found
+            if error.status.code != 404:
+                self.logger.error(f"Failed to delete CRD resources, with error: {error}")
+                delete_error = error
+        try:
+            delete_many(self.crd_resource_handler.lightkube_client, crd_resources_manifests)
+        except ApiError as error:
+            # do not log/report when resources were not found
+            if error.status.code != 404:
+                self.logger.error(f"Failed to delete K8s resources, with error: {error}")
+                delete_error = error
 
-    def _check_image_details(self):
+        if delete_error is not None:
+            raise delete_error
+
+        self.unit.status = MaintenanceStatus("K8s resources removed")
+
+    def _on_update_status(self, _) -> None:
+        """Handle update status event."""
+        self._on_event(_)
+        try:
+            self._check_status()
+        except ErrorWithStatus as err:
+            self.model.unit.status = err.status
+
+    def _on_event(self, event, force_conflicts: bool = False) -> None:
+        """Perform all required actions for the Charm.
+
+        Args:
+            force_conflicts (bool): Should only be used when need to resolved conflicts on K8s
+                                    resources.
+        """
         try:
-            image_details = self.image.fetch()
-        except OCIImageResourceError as e:
-            raise CheckFailed(f"{e.status.message}", e.status_type)
-        return image_details
+            self._check_container_connection()
+            self._check_leader()
+            self._apply_k8s_resources(force_conflicts=force_conflicts)
+            update_layer(
+                self._container_name,
+                self._container,
+                self._tensorboard_controller_layer,
+                self.logger,
+            )
+        except ErrorWithStatus as err:
+            self.model.unit.status = err.status
+            self.logger.error(f"Failed to handle {event} with error: {err}")
+            return
+
+        self.model.unit.status = ActiveStatus()
```

Comment:
After more thought I might change my mind, but my first reaction says instead of just setting Active maybe we should be using `_check_status()` here?

Reply:
Can't see why not!

Reply:
Note that if we end up using `_check_status` there we'll also need to mock pebble checks in our tests.

Reply:
Leaving this as is for now. It makes sense for `_check_status` to be managing the status, but in this case we have already covered the check part. I think we can confidently set that to Active given that `update_layer` has completed successfully.